Add fixed partitions

worker24h opened the pull request:
1. Add fixed partitions.  for example:
   ALTER TABLE test_table ADD PARTITION p0125 VALUES [("20190125"), ("20190126"));

ISSUE: #1990

morningman commented on fe/src/main/java/org/apache/doris/catalog/Replica.java:
```diff
@@ -111,7 +111,7 @@ public Replica(long replicaId, long backendId, int schemaHash, ReplicaState stat
     // for create tablet and restore
     public Replica(long replicaId, long backendId, ReplicaState state, long version, long versionHash, int schemaHash) {
-        this(replicaId, backendId, version, versionHash, schemaHash, -1, -1, state, -1L, 0L, version, versionHash);
+        this(replicaId, backendId, version, versionHash, schemaHash, 0L, -1, state, -1L, 0L, version, versionHash);
```
```suggestion
        this(replicaId, backendId, version, versionHash, schemaHash, 0L, 0L, state, -1L, 0L, version, versionHash);
```

morningman commented on fe/src/main/cup/sql_parser.cup:
```diff
@@ -1515,6 +1521,15 @@ partition_key_desc ::=
+fixed_partition_key_desc ::=
+    LBRACKET string_list:keys RPAREN
+    {:
+        if (keys.size() != 2) {
```
I think the proper format is:
`[("low1", "low2"), ("up1", "up2"))`

And should also support :
`[("low1", "low2", ..), ("up1"))`, and the default value should be filled with min value

Do we support MAXVALUE?

morningman reviewed (changes requested):
2 more things:
1. Add Documents
2. Unit Test

worker24h committed:
Add fixed partitions. for example:
ALTER TABLE test_table ADD PARTITION p0125 VALUES [("20190125"), ("20190126"));

morningman commented on fe/src/main/cup/sql_parser.cup:
```diff
@@ -1515,6 +1523,81 @@ partition_key_desc ::=
+        }
+        if (keys.get(0).toUpperCase().equals("MAXVALUE")) {
+            throw new Exception("the left value can't MAXVALUE.");
+        }
```
You should not check semantic here, this should be done in analyze phase in FE.
Here should only check syntax.

morningman commented on fe/src/main/cup/sql_parser.cup:
```diff
@@ -1515,6 +1523,81 @@ partition_key_desc ::=
+    :}
+    | LBRACKET multi_partition_keys_list:keys RPAREN
+      {:
+          if (keys.size() != 2) {
```
`keys.size() != 2` should already be guaranteed by the syntax your define here.

morningman commented on fe/src/main/java/org/apache/doris/analysis/LiteralExpr.java:
```diff
@@ -45,6 +45,9 @@ protected LiteralExpr(LiteralExpr other) {
     public static LiteralExpr create(String value, Type type) throws AnalysisException {
         Preconditions.checkArgument(type != Type.INVALID);
         LiteralExpr literalExpr = null;
+        if (value.toUpperCase().equals("MAXVALUE")) {
```
Can not do this!
`MAXVALUE` should be a keyword of syntex, NOT a keyword of LiteralExpr.
What if user create a string "MAXVALUE"?

morningman commented on fe/src/main/java/org/apache/doris/catalog/PartitionKey.java:
```diff
@@ -120,11 +120,34 @@ public boolean isMinValue() {
-            if (literalExpr != MaxLiteral.MAX_VALUE) {
-                return false;
+            if (literalExpr == MaxLiteral.MAX_VALUE) {
+                return true;
```
Only all sub keys are MAX_VALUE that this partition key can be MaxValue.
partition key like `(MAX_VALUE, "100")` is not max value.

morningman replied:
You need to modify this

worker24h replied:
ok

morningman commented on fe/src/main/java/org/apache/doris/catalog/RangePartitionInfo.java:
```diff
@@ -93,12 +93,60 @@ public void addPartition(long partitionId, Range<PartitionKey> range, DataProper
+                    PartitionKey.createPartitionKey(partKeyDesc.getUpperValues(), partitionColumns);
+
+            // check: [left, right), error if left equal right
+            if (lowKey.compareTo(upperKey) == 0) {
```
`lowKey.compareTo(upperKey) == 0` ==> `lowKey.compareTo(upperKey) >= 0` ?
If this is just for avoid `Range.closedOpen()`'s invalid check, you should add some comment here.

worker24h replied:
ok

morningman commented on fe/src/main/java/org/apache/doris/catalog/RangePartitionInfo.java:
```diff
@@ -93,12 +93,60 @@ public void addPartition(long partitionId, Range<PartitionKey> range, DataProper
+                }
+            }
+        } catch (AnalysisException e) {
+            throw new DdlException("Invalid range value format： " + e.getMessage());
```
```suggestion
            throw new DdlException("Invalid range value format: " + e.getMessage());
```

morningman commented on fe/src/main/java/org/apache/doris/catalog/RangePartitionInfo.java:
```diff
@@ -93,12 +93,60 @@ public void addPartition(long partitionId, Range<PartitionKey> range, DataProper
+            for (Map.Entry<Long, Range<PartitionKey>> entry : entries) {
+                Range<PartitionKey> nextRange = entry.getValue();
+                checkRangeIntersect(newRange, nextRange);
+                //keys only have one.
```
Why just check first key?

morningman commented on fe/src/main/java/org/apache/doris/analysis/PartitionValue.java:
```diff
@@ -0,0 +1,56 @@
+    public PartitionValue() {
+        isMaxValue = true;
+        value = null;
+    }
```
create a function like PartitionValue.createMax() to make it more explicit.

worker24h commented on fe/src/main/java/org/apache/doris/analysis/PartitionKeyDesc.java:
```diff
@@ -34,84 +34,121 @@
 public class PartitionKeyDesc implements Writable {
     // public static final PartitionKeyDesc MAX_VALUE = new PartitionKeyDesc();
     // lower values only be used for restore
```
ok

morningman reviewed (approved):
LGTM
